File: web/src/game/Shape.ts
import { V2 } from "./V2";

/**
 * Port of legacy/src/level/Shape.cpp. Parses the ASCII-art shape strings
 * used by addModel() in models.lua ("X" = occupied cell, "." = empty,
 * "\n" = next row).
 *
 * NOTE: width/height are the bounding box of the "X" marks only (the max
 * column/row that actually has an "X"), not the raw string's row length -
 * trailing "."-only padding on a row, or trailing "."-only rows, don't
 * count. This matches the original exactly (Shape::Shape only advances
 * max_x/max_y when it hits an 'X').
 */
export class Shape {
  readonly marks: readonly V2[];
  readonly w: number;
  readonly h: number;

  constructor(shape: string) {
    const marks: V2[] = [];
    let x = 0;
    let y = 0;
    let maxX = -1;
    let maxY = -1;

    for (const ch of shape) {
      switch (ch) {
        case "\n":
          y += 1;
          x = 0;
          break;
        case "X":
          marks.push(new V2(x, y));
          maxX = Math.max(maxX, x);
          maxY = Math.max(maxY, y);
          x += 1;
          break;
        case ".":
          x += 1;
          break;
        default:
          throw new Error(`bad shape char "${ch}" in shape "${shape}"`);
      }
    }

    this.marks = marks;
    this.w = maxX + 1;
    this.h = maxY + 1;
  }
}

File: solver/src/FishFillets.Physics/Shape.cs
namespace FishFillets.Physics;

/// <summary>
/// A model's occupied cells, relative to its anchor. Port of
/// legacy/src/level/Shape.cpp (web/src/game/Shape.ts): parses the ASCII-art
/// strings addModel() takes in models.lua - "X" occupied, "." empty, "\n" next
/// row.
///
/// NOTE: <see cref="W"/>/<see cref="H"/> are the bounding box of the "X" marks
/// only (the largest column/row that actually holds an "X"), not the raw
/// string's row length - trailing "."-padding doesn't count. Matches the
/// original exactly (Shape::Shape only advances max_x/max_y on an 'X').
///
/// Marks are stored once per model at level build and never re-parsed, so this
/// is the only place shape strings are touched.
/// </summary>
public sealed class Shape
{
    /// <summary>Cell offsets from the anchor, packed as (dx, dy) pairs.</summary>
    public readonly sbyte[] MarkX;

    public readonly sbyte[] MarkY;

    public readonly int W;

    public readonly int H;

    public Shape(string shape)
    {
        int count = 0;
        foreach (char c in shape)
        {
            if (c == 'X') count++;
        }

        MarkX = new sbyte[count];
        MarkY = new sbyte[count];

        int x = 0, y = 0, n = 0, maxX = -1, maxY = -1;
        foreach (char c in shape)
        {
            switch (c)
            {
                case '\n':
                    y++;
                    x = 0;
                    break;
                case 'X':
                    if (x > sbyte.MaxValue || y > sbyte.MaxValue)
                    {
                        throw new InvalidDataException($"shape mark out of sbyte range at ({x},{y})");
                    }

                    MarkX[n] = (sbyte)x;
                    MarkY[n] = (sbyte)y;
                    n++;
                    if (x > maxX) maxX = x;
                    if (y > maxY) maxY = y;
                    x++;
                    break;
                case '.':
                    x++;
                    break;
                default:
                    throw new InvalidDataException($"bad shape char '{c}'");
            }
        }

        W = maxX + 1;
        H = maxY + 1;
    }

    public int MarkCount => MarkX.Length;
}
